### PR TITLE
dnsdist: Fix several bugs in the TCP code path, add unit tests

### PR DESCRIPTION
**Pull request by rgacogne**
### Short description

This PR fixes several issues in corner cases, introduced during the refactoring for out-of-order support. It also adds a lot of unit tests for that TCP code, which was planned for 1.7.0 but fast-forwarded because of issues reported with 1.6.0-alpha1.

It also adds a new configuration option, `setTCPInternalPipeBufferSize`, that can be used to set the size of the buffer of the internal pipe used to pass queries to TCP worker threads, thus allowing a smoother handling of incoming TCP connections peaks.

The diff seems huge but most of that (3200 out of 3900 additions) comes from the new unit tests, which actually amount to more lines of code than the TCP code itself and provides a very nice code coverage (which unfortunately does not prevent bugs from still hiding in a unexpected state.. ;-)).

One caveat is that this drops source selection support other than via `SO_BINDTODEVICE`, because having to pass the source interface name and index down to the actual `sendmsg` code is very annoying. Perhaps we should at least look into supporting `IP_SENDIF` in addition to `SO_BINDTODEVICE`?

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [x] included documentation (including possible behaviour changes)
- [x] documented the code
- [ ] added or modified regression test(s)
- [x] added or modified unit test(s)

### CODE REVIEW

File: pdns/dnsdistdist/test-dnsdisttcp_cc.cc [hunk 1]
```diff
@@ -0,0 +1,3210 @@
...
+    BOOST_CHECK_EQUAL(s_writeBuffer.size(), 0U);
+    BOOST_CHECK_EQUAL(s_backendWriteBuffer.size(), query.size());
+    BOOST_CHECK(s_backendWriteBuffer == query);
+
+    /* we need to clear them now, otherwise we end up with dangling pointers to the steps via the TLS context, etc */
+    IncomingTCPConnectionState::clearAllDownstreamConnections();
+  }
+
+  {
+#if 0
```
Comment (omoerbeek) [inline pdns/dnsdistdist/test-dnsdisttcp_cc.cc:1551]: Intended to be switched off? If so, please add a comment why

**Comment by omoerbeek** [comment]
Nice to see more general TCP FastOpen support. I'm a bit worried about the source binding changes as you are using Linux specific functions. Both alternatives are not generally available.

**Comment by omoerbeek** [comment]
> Nice to see more general TCP FastOpen support. I'm a bit worried about the source binding changes as you are relying on specific functions. Both alternatives are not generally available.

Thinking about this a bit more, the old-fashioned way of binding to a source *address* before connect should still work, right?

File: pdns/tcpiohandler.hh [hunk 2]
```diff
@@ -197,35 +205,89 @@ public:
...
+
+#ifdef MSG_FASTOPEN
+    if (!d_conn && fastOpen) {
+      d_fastOpen = true;
+    }
+    else {
+      SConnectWithTimeout(d_socket, remote, /* no timeout, we will handle it ourselves */ 0);
+    }
+#else
+    SConnectWithTimeout(d_socket, d_ds->remote, /* no timeout, we will handle it ourselves */ 0);
```
Comment (omoerbeek) [inline pdns/tcpiohandler.hh:253]:
```suggestion
    SConnectWithTimeout(d_socket, remote, /* no timeout, we will handle it ourselves */ 0);
```

File: pdns/tcpiohandler.hh [hunk 3]
```diff
@@ -197,35 +205,89 @@ public:
...
+
+#ifdef MSG_FASTOPEN
+    if (!d_conn && fastOpen) {
+      d_fastOpen = true;
+    }
+    else {
+      SConnectWithTimeout(d_socket, remote, timeout);
+    }
+#else
+    SConnectWithTimeout(d_socket, d_ds->remote, timeout);
```
Comment (omoerbeek) [inline pdns/tcpiohandler.hh:285]:
```suggestion
    SConnectWithTimeout(d_socket, remote, timeout);
```

File: pdns/tcpiohandler.hh [hunk 4]
```diff
@@ -297,8 +359,24 @@ public:
       return d_conn->tryWrite(buffer, pos, toWrite);
     }
 
+    if (d_fastOpen) {
```
Comment (omoerbeek) [inline pdns/tcpiohandler.hh:362]:
```suggestion
#ifdef MSG_FASTOPEN
    if (d_fastOpen) {
```

File: pdns/tcpiohandler.hh [hunk 5]
```diff
@@ -297,8 +359,24 @@ public:
...
+        pos += sent;
+      }
+
+      if (pos < toWrite) {
+        return IOState::NeedWrite;
+      }
+
+      return IOState::Done;
+    }
+
```
Comment (omoerbeek) [inline pdns/tcpiohandler.hh:376]:
```suggestion
#endif /* MSG_FASTOPEN */
```

File: pdns/tcpiohandler.hh [hunk 6]
```diff
@@ -367,13 +445,14 @@ public:
...
+  bool getUnknownTicketKey() const
   {
     return d_conn && d_conn->getUnknownTicketKey();
   }
 
 private:
   std::unique_ptr<TLSConnection> d_conn{nullptr};
+  ComboAddress d_remote;
   int d_socket{-1};
   bool d_fastOpen{false};
```
Comment (omoerbeek) [inline pdns/tcpiohandler.hh:457]:
```suggestion
#if defined(MSG_FASTOPEN) || defined(TCP_FASTOPEN_CONNECT)
  bool d_fastOpen{false};
#endif
```

**Comment by rgacogne** [comment]
> Thinking about this a bit more, the old-fashioned way of binding to a source address before connect should still work, right?

Yes, that only concerns source **interface** selection, sorry!

File: pdns/tcpiohandler.hh [hunk 7]
```diff
@@ -106,6 +106,14 @@ protected:
 class TLSFrontend
```
Comment (omoerbeek) [inline pdns/tcpiohandler.hh:106]: Add `#include <netinet/tcp.h>` above, so we have the `TCP_FASTOPEN_CONNECT` connect if there.

**Comment by omoerbeek** [comment]
Something to investigate: I did not have `TCP_FASTOPEN_CONNECT`, but it seems I did not get the fallback to `MSG_FASTOPEN` either.

File: pdns/tcpiohandler.hh (hunk 6 above)
Comment (rgacogne) [reply pdns/tcpiohandler.hh:457]: We only use it for `MSG_FASTOPEN`, even. On it, thanks!

**Comment by rgacogne** [comment]
> Something to investigate: I did not have TCP_FASTOPEN_CONNECT, but it seems I did not get the fallback to MSG_FASTOPEN either.
> 
> The fallback to MSG_FASTOPEN if the include is not there is enabled, but does not seem tot work as intended for sdig using the write call.

Fixed the detection of `TCP_FASTOPEN_CONNECT` and the handling of `MSG_FASTOPEN` in blocking mode (only the non-blocking one was supported). I also added a `fastOpen` option to `sdig`.

File: pdns/dnsdistdist/test-dnsdisttcp_cc.cc (hunk 1 above)
Comment (rgacogne) [reply pdns/dnsdistdist/test-dnsdisttcp_cc.cc]: Oops, I only intended to disable them for a while when I was writing more tests. Fixing, thanks!

**Comment by omoerbeek** [comment]
NIce, when this is merged, I'll make a PR to change the two write calls in sdig into as single one.

**Review by omoerbeek — APPROVED** [review]
FastOpen works now as expected in sdig. I do see two writes in captures, but that is expected. I browsed the rest of the changes, and found nothing surprising.

**Comment by rgacogne** [comment]
Rebased to fix some commits not compiling.
